xapian-core/tests/unittest.cc

```diff
@@ -680,6 +682,86 @@ static bool test_uuid1()
     return true;
 }
 
+// Classes used by movesupport1 test
+class A : public Xapian::Internal::intrusive_base {
+    int x = 0;
+  public:
+    A(int x_) : x(x_) {}
+    int get_x() const {
+	return x;
+    }
+};
+
+class B : public Xapian::Internal::opt_intrusive_base {
+    int x = 0;
+  public:
+    B(int x_) : x(x_) {}
+    int get_x() const {
+	return x;
+    }
+    B * release() {
+	opt_intrusive_base::release();
+	return this;
+    }
+};
+
+static bool test_movesupport1()
+{
+    {
+	// Test move semantics support for intrusive_ptr class
+	Xapian::Internal::intrusive_ptr<A> p1(new A{5});
+	Xapian::Internal::intrusive_ptr<A> p3;
+
+	// Test move constructor
+	Xapian::Internal::intrusive_ptr<A> p2(std::move(p1));
+	TEST_EQUAL(p2->get_x(), 5);
+	TEST_EQUAL(p1.get(), 0);
+
+	// Test move assignment
+	p3 = std::move(p2);
+	TEST_EQUAL(p3->get_x(), 5);
+	TEST_EQUAL(p2.get(), 0);
+    }
+
+
+    {
+	// Same test for intrusive_ptr_nonnull class
+	Xapian::Internal::intrusive_ptr_nonnull<A> p1(new A{5});
+	Xapian::Internal::intrusive_ptr_nonnull<A> p3(new A{6});
+
+	// Test move constructor
+	Xapian::Internal::intrusive_ptr_nonnull<A> p2(std::move(p1));
+	TEST_EQUAL(p2->get_x(), 5);
+	TEST_EQUAL(p1.get(), 0);
+
+	// Test move assignment
+	p3 = std::move(p2);
+	TEST_EQUAL(p3->get_x(), 5);
+	TEST_EQUAL(p2.get(), 0);
+    }
+
+    {
+	// Same test for opt_intrusive_ptr class
+	B * b1 = new B{5};
+	b1->release();
+	Xapian::Internal::opt_intrusive_ptr<B> p1(b1);
+	Xapian::Internal::opt_intrusive_ptr<B> p3;
+
+	// Test move constructor
+	Xapian::Internal::opt_intrusive_ptr<B> p2(std::move(p1));
+	TEST_EQUAL(p2->get_x(), 5);
+	TEST_EQUAL(p2->_refs, 2);
```

Comment:
The `_refs` member is really an implementation detail, so it's violating the abstraction to look at it here.  It's only `public` at all for convenience of implementation.

If it is useful to inspect, I'd suggest that instead you do `auto before_refs = p1->_refs` before attempting to invoke the move constructor, and then `TEST_EQUAL(p2->_refs, before_refs);` here to test that the reference count hasn't changed.  That at least avoids encoding knowledge here about what the value means exactly.

But I'd think you could have `B`'s constructor take a `bool&` parameter which its destructor sets to false, and use that to test that the object hasn't been deleted at the end of this block, but has been just after this block.  That should ensure that we catch any bugs in updating the reference count in the move constructor or move assignment.

Reply:
> But I'd think you could have B's constructor take a bool& parameter which its destructor sets to false,
I followed this. It's better. Removed check for _refs value.